### PR TITLE
HttpServer: Make read timeout configurable but disabled by default

### LINKED ISSUE #31440: Security: Add timeout options for main http server
security scan report a moderate security problem: “Slow HTTP denial of service attack“(slow header,slow body,slow read), I see http server code(/pkg/api/http_server.go)
	hs.httpSrv = &http.Server{
		Addr:    net.JoinHostPort(host, setting.HttpPort),
		Handler: hs.macaron,
	}
it use default values，so I can't avoid this problem by configuration
can you add several timeout parameters for http server configuration control？example parameters：ReadHeaderTimeout、ReadTimeout、WriteTimeout、IdleTimeout、MaxHeaderBytes

### PR DESCRIPTION
**Pull request by bergquist**
Closes https://github.com/grafana/grafana/issues/31440

### CODE REVIEW

**Comment by torkelo** [comment]
This should have a changelog entry, updated title to be more user describing. The question is should this also be mentioned under "breaking change" in the changelog/release notes?

**Review by aknuds1 — COMMENTED** [review]
Just have a couple of questions:

1. Should we introduce a default timeout (breaking change)?
2. Should we document how to not have a timeout (guessing this is by providing a 0 duration)?

**Comment by torkelo** [comment]
Ah, good point, needs update to configuration

File: docs/sources/administration/configuration.md [hunk 1]
```diff
@@ -268,6 +268,11 @@ Specify a full HTTP URL address to the root of your Grafana CDN assets. Grafana
 For example, given a cdn url like `https://cdn.myserver.com` grafana will try to load a javascript file from
 `http://cdn.myserver.com/grafana-oss/v7.4.0/public/build/app.<hash>.js`.
 
+### read_timeout
+
+Maximum duration before timing out read of the incoming request, and closing idle connections.
```
Comment (achatterjee-grafana) [inline docs/sources/administration/configuration.md:273]:
```suggestion
Sets the maximum time in minutes before timing out read of an incoming request and closing idle connections. The default is five (5) minutes.
```

**Review by achatterjee-grafana — COMMENTED** [review]
Added copy-edit suggestion.

**Comment by bergquist** [comment]
Switched to disable this by default. I think we might want to use the ReadHeadersTimeout instead and deal with timeouts in the handlers instead. I'm not sure but this will enable operators to set a read timeout if they want and we can explore how we want to deal with timeouts later on without changing the default value.

**Review by aknuds1 — APPROVED** [review]
LGTM

**Comment by oddlittlebird** [comment]
This needs a milestone.

**Comment by bergquist** [comment]
@oddlittlebird is that really required for a small change like this? I would think being part of the changelog would be enough.

**Comment by oddlittlebird** [comment]
https://grafana.com/docs/grafana/latest/whatsnew/whats-new-in-v7-5/

**Comment by bergquist** [comment]
Added to what new and updates the help text in config files

File: docs/sources/administration/configuration.md [hunk 2]
```diff
@@ -268,6 +268,11 @@ Specify a full HTTP URL address to the root of your Grafana CDN assets. Grafana
 For example, given a cdn url like `https://cdn.myserver.com` grafana will try to load a javascript file from
 `http://cdn.myserver.com/grafana-oss/v7.4.0/public/build/app.<hash>.js`.
 
+### read_timeout
+
+Sets the maximum time using a duration format (5s/5m/5ms) before timing out read of an incoming request and closing idle connections.
+`0` means there is no timeout for reading the request
```
Comment (achatterjee-grafana) [inline docs/sources/administration/configuration.md:274]:
```suggestion
`0` means there is no timeout for reading the request.
```

File: conf/sample.ini [hunk 3]
```diff
@@ -70,6 +70,10 @@
 # CDN Url
 ;cdn_url =
 
+# Sets the maximum time using a duration format (5s/5m/5ms) before timing out read of an incoming request and closing idle connections.
+# `0` means there is no timeout for reading the request
```
Comment (achatterjee-grafana) [inline conf/sample.ini:74]:
```suggestion
# `0` means there is no timeout for reading the request.
```

File: conf/defaults.ini [hunk 4]
```diff
@@ -69,6 +69,10 @@ socket = /tmp/grafana.sock
 # CDN Url
 cdn_url =
 
+# Sets the maximum time in minutes before timing out read of an incoming request and closing idle connections. 
+# `0` means there is no timeout for reading the request
```
Comment (achatterjee-grafana) [inline conf/defaults.ini:73]:
```suggestion
# `0` means there is no timeout for reading the request.
```

File: docs/sources/whatsnew/whats-new-in-v7-5.md [hunk 5]
```diff
@@ -105,3 +105,13 @@ If you enable the feature, then you can use template variables in reports.
...
+
+## Updated configuration
+
+```
+[server]
+read_timeout = 0
+```
+
+Sets the maximum time using a duration format (5s/5m/5ms) before timing out read of an incoming request and closing idle connections.
+`0` means there is no timeout for reading the request
```
Comment (achatterjee-grafana) [inline docs/sources/whatsnew/whats-new-in-v7-5.md:117]:
```suggestion
`0` means there is no timeout for reading the request.
```

**Review by achatterjee-grafana — APPROVED** [review]
Added a missing period at the end of a few sentences.

**Pull request merged**